### DISCUSSION TITLE
General reuse of website components

### DISCUSSION THREAD

**Discussion by z3dev** [jscad/OpenJSCAD.org#848] [General]
There's been some requests to split out some of the components found in the core/web packages. The requested components are for:
- parsing parameters
- creating forms for the parameters
- processing parameters

There may be other common components as well. Such as those found in the core package.
- creating fake file systems, which are passed into evaluation() functions

So far, the requests are for components that can be used by 'generic' JavaScript applications.

**Comment by hrgdavor** [comment]
I have alternative version of those 3 parameters related libs with some added features. It has no dependencies.
I have made an additional split to allow myself to generate form manually that has name="xxx" matching attribute on inputs.
This way I can have benefits of parameter processing with a custom form.

Viewer component should support all of the functionalities that are on openjscad web but without UI elements.
UI elements that provide those options should be separate.

Viewer options
 - show grid
 - show axis
 - corner axis preview
 - theme - name of builtin theme, or object with theme params(colors, and maybe outline in the future)
 - auto rotate
 - auto reload
 - zoomToFit - after script or paramteres change
 - allow drag/drop - in where it is embeded viewer that should not allow drag&drop
 - getCameraPosition
 - setCameraPosition
 
Viewer methods
 - zoomToFit  - if user wants to manually call this
 - getParameterDefinitions - viewer will have access to script, but parameters form should be separate component
 - setFs  - set fake file system for loading additional things by the executed script
 - executeScript - script code
 - loadScript  - url or reference to file from upload component or drag&drop

Viewer events
 - parameter definitions loaded / changed
 - model calculated / updated
